test/validation/api/ml/ml.c

```diff
@@ -0,0 +1,577 @@
+
+		CU_ASSERT(compl_evs[i] != ODP_ML_COMPL_INVALID);
+
+		if (compl_evs[i] == ODP_ML_COMPL_INVALID)
+			break;
+
+		uarea = odp_ml_compl_user_area(compl_evs[i]);
+
+		CU_ASSERT(*uarea == UAREA);
+	}
+
+	for (uint32_t j = 0; j < i; j++)
+		odp_ml_compl_free(compl_evs[j]);
+
+	odp_pool_destroy(pool);
+}
+
+static void test_ml_fp32_to_uint8(void)
+{
+	uint8_t u8[8];
+	float fp[8] = {-20.f, -16.4f, -14.6f, -12.5f, 0, 31.4f, 80.f, 96.3f};
+	uint8_t expected[8] = {0, 0, 4, 10, 43, 127, 255, 255};
+	float scale = 0.3746f;
+	uint8_t zero_point = 43;
+
+	odp_ml_fp32_to_uint8(u8, fp, 8, scale, zero_point);
+	for (uint32_t i = 0; i < 8; i++)
+		CU_ASSERT_EQUAL(u8[i], expected[i]);
+}
+
+static void test_ml_fp32_from_uint8(void)
+{
+	float fp[4];
+	float scale = 0.4f;
+	uint8_t zero_point = 43;
+	uint8_t u8[4] = {0, 43, 145, 255};
+	float expected[4] = {-17.2f, 0.0f, 40.8f, 84.8f};
+
+	odp_ml_fp32_from_uint8(fp, u8, 4, scale, zero_point);
+	for (uint32_t i = 0; i < 4; i++)
+		CU_ASSERT_EQUAL(fp[i], expected[i]);
+}
+
+static void test_ml_fp32_to_int8(void)
+{
+	int8_t i8[5];
+	float scale = 0.0223f;
+	int8_t zero_point = 0;
+	float fp32[5] = {-3.4f, -2.5f, 0, 1.4f, 2.9f};
+	int8_t i8_expected[5] = {-127, -112, 0, 63, 127};
+
+	odp_ml_fp32_to_int8(i8, fp32, 5, scale, zero_point);
+
+	for (uint32_t i = 0; i < 5; i++)
+		CU_ASSERT_EQUAL(i8[i], i8_expected[i]);
+}
+
+static void test_ml_fp32_to_int8_positive_zp(void)
+{
+	int8_t i8[6];
+	float scale = 0.0223f;
+	int8_t zero_point = 56;
+	float fp32[6] = {-4.1f, -3.4f, -2.5f, 0, 1.4f, 2.9f};
+	int8_t i8_expected[6] = {-127, -96, -56, 56, 119, 127};
+
+	odp_ml_fp32_to_int8(i8, fp32, 6, scale, zero_point);
+
+	for (uint32_t i = 0; i < 6; i++)
+		CU_ASSERT_EQUAL(i8[i], i8_expected[i]);
+}
+
+static void test_ml_fp32_to_int8_negative_zp(void)
+{
+	int8_t i8[6];
+	float scale = 0.0223f;
+	int8_t zero_point = -56;
+	float fp32[6] = {-3.4f, -2.5f, 0, 1.4f, 2.9f, 4.1f};
+	int8_t i8_expected[6] = {-127, -127, -56, 7, 74, 127};
+
+	odp_ml_fp32_to_int8(i8, fp32, 6, scale, zero_point);
+
+	for (uint32_t i = 0; i < 6; i++)
+		CU_ASSERT_EQUAL(i8[i], i8_expected[i]);
+}
+
+static void test_ml_fp32_from_int8(void)
+{
+	float fp32[6];
+	float scale = 0.05f;
+	int8_t zero_point = 56;
+	int8_t i8[6] = {-128, 46, 0, 56, 85, 127};
+	float fp32_expected[6] = {-9.2f, -0.5f, -2.8f, 0.0f, 1.45f, 3.55f};
+
+	odp_ml_fp32_from_int8(fp32, i8, 6, scale, zero_point);
+
+	for (uint32_t i = 0; i < 6; i++)
+		CU_ASSERT_EQUAL(fp32[i], fp32_expected[i]);
+}
+
+static int approx_equal(double a, double b)
+{
+	const double tolerance = .01;
+
+	if (a < 0 && b < 0) {
+		a = -a;
+		b = -b;
+	}
+
+	if (a < b) {
+		double tmp = a;
+
+		a = b;
+		b = tmp;
+	}
+
+	return (a * (1 - tolerance) < b && a * (1 + tolerance) > b);
+}
+
+static void test_ml_fp32_fp16(void)
+{
+	float fp32[4];
+	uint16_t fp16[4];
+	float big = 1, small = 1;
+	const float phi = 1.618033988749;
+
+	fp32[0] = 0.0;
+	fp32[1] = -0.0;
+	fp32[2] = 1.0 / 0.0;
+	fp32[3] = 0.0 / 0.0;
```

Comment:
I'd drop testing NaN values as they may be problematic for some implementations (e.g. Arm alternative half-precision).

Reply:
Division by zero has undefined behavior in C, so it would make sense not to do it.